// @workspace/api-zod/src/generated/api.ts
/**
 * Generated by orval v8.20.0 🍺
 * Do not edit manually.
 * Api
 * API specification
 * OpenAPI spec version: 0.1.0
 */
import * as zod from 'zod';


/**
 * Returns server health status
 * @summary Health check
 */
export const HealthCheckResponse = zod.object({
  "status": zod.string()
})


/**
 * @summary Request a one-time login code for a phone number
 */
export const requestOtpBodyPhoneMin = 6;
export const requestOtpBodyPhoneMax = 15;



export const RequestOtpBody = zod.object({
  "phone": zod.string().min(requestOtpBodyPhoneMin).max(requestOtpBodyPhoneMax)
})

export const RequestOtpResponse = zod.object({
  "phone": zod.string(),
  "otp": zod.string().describe('Simulated OTP, shown directly since no real SMS gateway is connected in this demo.'),
  "expiresInSeconds": zod.number()
})


/**
 * @summary Verify the one-time code and start a session
 */
export const VerifyOtpBody = zod.object({
  "phone": zod.string(),
  "otp": zod.string(),
  "name": zod.string().nullish().describe('Optional display name to set on first login.')
})

export const VerifyOtpResponse = zod.object({
  "user": zod.object({
  "id": zod.number(),
  "phone": zod.string(),
  "name": zod.string(),
  "createdAt": zod.coerce.date()
}),
  "wallet": zod.object({
  "balanceCents": zod.number()
})
})


/**
 * @summary Get the current logged-in user and wallet
 */
export const GetCurrentSessionResponse = zod.object({
  "user": zod.object({
  "id": zod.number(),
  "phone": zod.string(),
  "name": zod.string(),
  "createdAt": zod.coerce.date()
}),
  "wallet": zod.object({
  "balanceCents": zod.number()
})
})


/**
 * @summary End the current session
 */
export const LogoutResponse = zod.void()


/**
 * @summary Get the current user's wallet balance
 */
export const GetWalletResponse = zod.object({
  "balanceCents": zod.number()
})


/**
 * @summary Simulate adding money to the wallet
 */
export const addMoneyBodyAmountCentsMin = 100;



export const AddMoneyBody = zod.object({
  "amountCents": zod.number().min(addMoneyBodyAmountCentsMin),
  "method": zod.enum(['upi', 'debit_card', 'credit_card', 'netbanking'])
})

export const AddMoneyResponse = zod.object({
  "wallet": zod.object({
  "balanceCents": zod.number()
}),
  "transaction": zod.object({
  "id": zod.number(),
  "type": zod.enum(['add_money', 'transfer_out', 'transfer_in', 'recharge', 'bill_payment']),
  "amountCents": zod.number(),
  "status": zod.string(),
  "counterpartyName": zod.string().nullish(),
  "counterpartyPhone": zod.string().nullish(),
  "category": zod.string().nullish(),
  "note": zod.string().nullish(),
  "createdAt": zod.coerce.date()
})
})


/**
 * @summary List the current user's transactions, most recent first
 */
export const listTransactionsQueryLimitDefault = 50;
export const listTransactionsQueryLimitMax = 200;



export const ListTransactionsQueryParams = zod.object({
  "limit": zod.coerce.number().min(1).max(listTransactionsQueryLimitMax).default(listTransactionsQueryLimitDefault)
})

export const ListTransactionsResponseItem = zod.object({
  "id": zod.number(),
  "type": zod.enum(['add_money', 'transfer_out', 'transfer_in', 'recharge', 'bill_payment']),
  "amountCents": zod.number(),
  "status": zod.string(),
  "counterpartyName": zod.string().nullish(),
  "counterpartyPhone": zod.string().nullish(),
  "category": zod.string().nullish(),
  "note": zod.string().nullish(),
  "createdAt": zod.coerce.date()
})
export const ListTransactionsResponse = zod.array(ListTransactionsResponseItem)


/**
 * @summary Get a single transaction by id
 */
export const GetTransactionParams = zod.object({
  "id": zod.coerce.number()
})

export const GetTransactionResponse = zod.object({
  "id": zod.number(),
  "type": zod.enum(['add_money', 'transfer_out', 'transfer_in', 'recharge', 'bill_payment']),
  "amountCents": zod.number(),
  "status": zod.string(),
  "counterpartyName": zod.string().nullish(),
  "counterpartyPhone": zod.string().nullish(),
  "category": zod.string().nullish(),
  "note": zod.string().nullish(),
  "createdAt": zod.coerce.date()
})


/**
 * @summary Send money to another user by phone number
 */
export const createTransferBodyAmountCentsMin = 100;



export const CreateTransferBody = zod.object({
  "toPhone": zod.string(),
  "amountCents": zod.number().min(createTransferBodyAmountCentsMin),
  "note": zod.string().nullish()
})

export const CreateTransferResponse = zod.object({
  "wallet": zod.object({
  "balanceCents": zod.number()
}),
  "transaction": zod.object({
  "id": zod.number(),
  "type": zod.enum(['add_money', 'transfer_out', 'transfer_in', 'recharge', 'bill_payment']),
  "amountCents": zod.number(),
  "status": zod.string(),
  "counterpartyName": zod.string().nullish(),
  "counterpartyPhone": zod.string().nullish(),
  "category": zod.string().nullish(),
  "note": zod.string().nullish(),
  "createdAt": zod.coerce.date()
})
})


/**
 * @summary Look up a user's public profile by phone number
 */
export const LookupUserQueryParams = zod.object({
  "phone": zod.coerce.string()
})

export const LookupUserResponse = zod.object({
  "name": zod.string(),
  "phone": zod.string()
})


/**
 * @summary List saved payees for the current user
 */
export const ListContactsResponseItem = zod.object({
  "id": zod.number(),
  "name": zod.string(),
  "phone": zod.string(),
  "nickname": zod.string().nullish(),
  "createdAt": zod.coerce.date()
})
export const ListContactsResponse = zod.array(ListContactsResponseItem)


/**
 * @summary Save a new payee
 */

export const createContactBodyPhoneMin = 6;



export const CreateContactBody = zod.object({
  "name": zod.string().min(1),
  "phone": zod.string().min(createContactBodyPhoneMin),
  "nickname": zod.string().nullish()
})

export const CreateContactResponse = zod.object({
  "id": zod.number(),
  "name": zod.string(),
  "phone": zod.string(),
  "nickname": zod.string().nullish(),
  "createdAt": zod.coerce.date()
})


/**
 * @summary Remove a saved payee
 */
export const DeleteContactParams = zod.object({
  "id": zod.coerce.number()
})

export const DeleteContactResponse = zod.void()


/**
 * @summary List available recharge and bill payment providers
 */
export const ListBillersQueryParams = zod.object({
  "category": zod.enum(['mobile_recharge', 'dth', 'electricity', 'water', 'gas', 'broadband', 'insurance', 'loan_payment']).optional()
})

export const ListBillersResponseItem = zod.object({
  "id": zod.string(),
  "category": zod.enum(['mobile_recharge', 'dth', 'electricity', 'water', 'gas', 'broadband', 'insurance', 'loan_payment']),
  "name": zod.string()
})
export const ListBillersResponse = zod.array(ListBillersResponseItem)


/**
 * @summary Pay a recharge or bill using a biller
 */

export const createRechargeBodyAmountCentsMin = 100;



export const CreateRechargeBody = zod.object({
  "billerId": zod.string(),
  "accountNumber": zod.string().min(1),
  "amountCents": zod.number().min(createRechargeBodyAmountCentsMin)
})

export const CreateRechargeResponse = zod.object({
  "wallet": zod.object({
  "balanceCents": zod.number()
}),
  "transaction": zod.object({
  "id": zod.number(),
  "type": zod.enum(['add_money', 'transfer_out', 'transfer_in', 'recharge', 'bill_payment']),
  "amountCents": zod.number(),
  "status": zod.string(),
  "counterpartyName": zod.string().nullish(),
  "counterpartyPhone": zod.string().nullish(),
  "category": zod.string().nullish(),
  "note": zod.string().nullish(),
  "createdAt": zod.coerce.date()
})
})


/**
 * @summary Aggregated balance, recent activity, and spend breakdown
 */
export const GetDashboardSummaryResponse = zod.object({
  "balanceCents": zod.number(),
  "totalAddedCents": zod.number(),
  "totalSpentCents": zod.number(),
  "recentTransactions": zod.array(zod.object({
  "id": zod.number(),
  "type": zod.enum(['add_money', 'transfer_out', 'transfer_in', 'recharge', 'bill_payment']),
  "amountCents": zod.number(),
  "status": zod.string(),
  "counterpartyName": zod.string().nullish(),
  "counterpartyPhone": zod.string().nullish(),
  "category": zod.string().nullish(),
  "note": zod.string().nullish(),
  "createdAt": zod.coerce.date()
})),
  "spendByCategory": zod.array(zod.object({
  "category": zod.string(),
  "totalCents": zod.number()
}))
})
