// @workspace/api-client-react/src/generated/api.ts
/**
 * Generated by orval v8.20.0 🍺
 * Do not edit manually.
 * Api
 * API specification
 * OpenAPI spec version: 0.1.0
 */
import {
  useMutation,
  useQuery
} from '@tanstack/react-query';
import type {
  MutationFunction,
  QueryFunction,
  QueryKey,
  UseMutationOptions,
  UseMutationResult,
  UseQueryOptions,
  UseQueryResult
} from '@tanstack/react-query';

import type {
  AddMoneyInput,
  Biller,
  Contact,
  ContactInput,
  DashboardSummary,
  ErrorResponse,
  HealthStatus,
  ListBillersParams,
  ListTransactionsParams,
  LookupUserParams,
  OtpRequestInput,
  OtpRequestResult,
  OtpVerifyInput,
  PublicUser,
  RechargeInput,
  SessionInfo,
  Transaction,
  TransferInput,
  Wallet,
  WalletMutationResult
} from './api.schemas';

import { customFetch } from '../custom-fetch';
import type { ErrorType , BodyType } from '../custom-fetch';

type AwaitedInput<T> = PromiseLike<T> | T;

      type Awaited<O> = O extends AwaitedInput<infer T> ? T : never;


type SecondParameter<T extends (...args: never) => unknown> = Parameters<T>[1];



const withQueryKey = <T extends object, K>(query: T, queryKey: K): T & { queryKey: K } => {
  const result = { queryKey } as T & { queryKey: K };
  for (const key of Object.keys(query)) {
    // The explicit queryKey always wins, matching the previous
    // `{ ...query, queryKey }` spread where it was set last.
    if (key === 'queryKey') continue;
    Object.defineProperty(result, key, {
      enumerable: true,
      configurable: true,
      get: () => (query as Record<string, unknown>)[key],
    });
  }
  return result;
};

export const getHealthCheckUrl = () => {




  return `/api/healthz`
}

/**
 * Returns server health status
 * @summary Health check
 */
export const healthCheck = async ( options?: RequestInit): Promise<HealthStatus> => {

  return customFetch<HealthStatus>(getHealthCheckUrl(),
  {
    ...options,
    method: 'GET'


  }
);}





export const getHealthCheckQueryKey = () => {
    return [
    `/api/healthz`
    ] as const;
    }


export const getHealthCheckQueryOptions = <TData = Awaited<ReturnType<typeof healthCheck>>, TError = ErrorType<unknown>>( options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof healthCheck>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getHealthCheckQueryKey();



    const queryFn: QueryFunction<Awaited<ReturnType<typeof healthCheck>>> = ({ signal }) => healthCheck({ signal, ...requestOptions });





   return  { queryKey, queryFn, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof healthCheck>>, TError, TData> & { queryKey: QueryKey }
}

export type HealthCheckQueryResult = NonNullable<Awaited<ReturnType<typeof healthCheck>>>
export type HealthCheckQueryError = ErrorType<unknown>


/**
 * @summary Health check
 */

export function useHealthCheck<TData = Awaited<ReturnType<typeof healthCheck>>, TError = ErrorType<unknown>>(
  options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof healthCheck>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getHealthCheckQueryOptions(options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getRequestOtpUrl = () => {




  return `/api/auth/request-otp`
}

/**
 * @summary Request a one-time login code for a phone number
 */
export const requestOtp = async (otpRequestInput: OtpRequestInput, options?: RequestInit): Promise<OtpRequestResult> => {

  return customFetch<OtpRequestResult>(getRequestOtpUrl(),
  {
    ...options,
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(otpRequestInput)
  }
);}





export const getRequestOtpMutationOptions = <TError = ErrorType<ErrorResponse>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof requestOtp>>, TError,{data: BodyType<OtpRequestInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof requestOtp>>, TError,{data: BodyType<OtpRequestInput>}, TContext> => {

const mutationKey = ['requestOtp'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof requestOtp>>, {data: BodyType<OtpRequestInput>}> = (props) => {
          const {data} = props ?? {};

          return  requestOtp(data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type RequestOtpMutationResult = NonNullable<Awaited<ReturnType<typeof requestOtp>>>
    export type RequestOtpMutationBody = BodyType<OtpRequestInput>
    export type RequestOtpMutationError = ErrorType<ErrorResponse>

    /**
 * @summary Request a one-time login code for a phone number
 */
export const useRequestOtp = <TError = ErrorType<ErrorResponse>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof requestOtp>>, TError,{data: BodyType<OtpRequestInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof requestOtp>>,
        TError,
        {data: BodyType<OtpRequestInput>},
        TContext
      > => {
      return useMutation(getRequestOtpMutationOptions(options));
    }

export const getVerifyOtpUrl = () => {




  return `/api/auth/verify-otp`
}

/**
 * @summary Verify the one-time code and start a session
 */
export const verifyOtp = async (otpVerifyInput: OtpVerifyInput, options?: RequestInit): Promise<SessionInfo> => {

  return customFetch<SessionInfo>(getVerifyOtpUrl(),
  {
    ...options,
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(otpVerifyInput)
  }
);}





export const getVerifyOtpMutationOptions = <TError = ErrorType<ErrorResponse>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof verifyOtp>>, TError,{data: BodyType<OtpVerifyInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof verifyOtp>>, TError,{data: BodyType<OtpVerifyInput>}, TContext> => {

const mutationKey = ['verifyOtp'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof verifyOtp>>, {data: BodyType<OtpVerifyInput>}> = (props) => {
          const {data} = props ?? {};

          return  verifyOtp(data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type VerifyOtpMutationResult = NonNullable<Awaited<ReturnType<typeof verifyOtp>>>
    export type VerifyOtpMutationBody = BodyType<OtpVerifyInput>
    export type VerifyOtpMutationError = ErrorType<ErrorResponse>

    /**
 * @summary Verify the one-time code and start a session
 */
export const useVerifyOtp = <TError = ErrorType<ErrorResponse>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof verifyOtp>>, TError,{data: BodyType<OtpVerifyInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof verifyOtp>>,
        TError,
        {data: BodyType<OtpVerifyInput>},
        TContext
      > => {
      return useMutation(getVerifyOtpMutationOptions(options));
    }

export const getGetCurrentSessionUrl = () => {




  return `/api/auth/me`
}

/**
 * @summary Get the current logged-in user and wallet
 */
export const getCurrentSession = async ( options?: RequestInit): Promise<SessionInfo> => {

  return customFetch<SessionInfo>(getGetCurrentSessionUrl(),
  {
    ...options,
    method: 'GET'


  }
);}





export const getGetCurrentSessionQueryKey = () => {
    return [
    `/api/auth/me`
    ] as const;
    }


export const getGetCurrentSessionQueryOptions = <TData = Awaited<ReturnType<typeof getCurrentSession>>, TError = ErrorType<ErrorResponse>>( options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof getCurrentSession>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getGetCurrentSessionQueryKey();



    const queryFn: QueryFunction<Awaited<ReturnType<typeof getCurrentSession>>> = ({ signal }) => getCurrentSession({ signal, ...requestOptions });





   return  { queryKey, queryFn, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof getCurrentSession>>, TError, TData> & { queryKey: QueryKey }
}

export type GetCurrentSessionQueryResult = NonNullable<Awaited<ReturnType<typeof getCurrentSession>>>
export type GetCurrentSessionQueryError = ErrorType<ErrorResponse>


/**
 * @summary Get the current logged-in user and wallet
 */

export function useGetCurrentSession<TData = Awaited<ReturnType<typeof getCurrentSession>>, TError = ErrorType<ErrorResponse>>(
  options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof getCurrentSession>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getGetCurrentSessionQueryOptions(options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getLogoutUrl = () => {




  return `/api/auth/logout`
}

/**
 * @summary End the current session
 */
export const logout = async ( options?: RequestInit): Promise<void> => {

  return customFetch<void>(getLogoutUrl(),
  {
    ...options,
    method: 'POST'


  }
);}





export const getLogoutMutationOptions = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof logout>>, TError,void, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof logout>>, TError,void, TContext> => {

const mutationKey = ['logout'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof logout>>, void> = () => {


          return  logout(requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type LogoutMutationResult = NonNullable<Awaited<ReturnType<typeof logout>>>

    export type LogoutMutationError = ErrorType<unknown>

    /**
 * @summary End the current session
 */
export const useLogout = <TError = ErrorType<unknown>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof logout>>, TError,void, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof logout>>,
        TError,
        void,
        TContext
      > => {
      return useMutation(getLogoutMutationOptions(options));
    }

export const getGetWalletUrl = () => {




  return `/api/wallet`
}

/**
 * @summary Get the current user's wallet balance
 */
export const getWallet = async ( options?: RequestInit): Promise<Wallet> => {

  return customFetch<Wallet>(getGetWalletUrl(),
  {
    ...options,
    method: 'GET'


  }
);}





export const getGetWalletQueryKey = () => {
    return [
    `/api/wallet`
    ] as const;
    }


export const getGetWalletQueryOptions = <TData = Awaited<ReturnType<typeof getWallet>>, TError = ErrorType<ErrorResponse>>( options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof getWallet>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getGetWalletQueryKey();



    const queryFn: QueryFunction<Awaited<ReturnType<typeof getWallet>>> = ({ signal }) => getWallet({ signal, ...requestOptions });





   return  { queryKey, queryFn, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof getWallet>>, TError, TData> & { queryKey: QueryKey }
}

export type GetWalletQueryResult = NonNullable<Awaited<ReturnType<typeof getWallet>>>
export type GetWalletQueryError = ErrorType<ErrorResponse>


/**
 * @summary Get the current user's wallet balance
 */

export function useGetWallet<TData = Awaited<ReturnType<typeof getWallet>>, TError = ErrorType<ErrorResponse>>(
  options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof getWallet>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getGetWalletQueryOptions(options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getAddMoneyUrl = () => {




  return `/api/wallet/add-money`
}

/**
 * @summary Simulate adding money to the wallet
 */
export const addMoney = async (addMoneyInput: AddMoneyInput, options?: RequestInit): Promise<WalletMutationResult> => {

  return customFetch<WalletMutationResult>(getAddMoneyUrl(),
  {
    ...options,
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(addMoneyInput)
  }
);}





export const getAddMoneyMutationOptions = <TError = ErrorType<ErrorResponse>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof addMoney>>, TError,{data: BodyType<AddMoneyInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof addMoney>>, TError,{data: BodyType<AddMoneyInput>}, TContext> => {

const mutationKey = ['addMoney'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof addMoney>>, {data: BodyType<AddMoneyInput>}> = (props) => {
          const {data} = props ?? {};

          return  addMoney(data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type AddMoneyMutationResult = NonNullable<Awaited<ReturnType<typeof addMoney>>>
    export type AddMoneyMutationBody = BodyType<AddMoneyInput>
    export type AddMoneyMutationError = ErrorType<ErrorResponse>

    /**
 * @summary Simulate adding money to the wallet
 */
export const useAddMoney = <TError = ErrorType<ErrorResponse>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof addMoney>>, TError,{data: BodyType<AddMoneyInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof addMoney>>,
        TError,
        {data: BodyType<AddMoneyInput>},
        TContext
      > => {
      return useMutation(getAddMoneyMutationOptions(options));
    }

export const getListTransactionsUrl = (params?: ListTransactionsParams,) => {
  const normalizedParams = new URLSearchParams();

  Object.entries(params || {}).forEach(([key, value]) => {

    if (value !== undefined) {
      normalizedParams.append(key, value === null ? 'null' : String(value))
    }
  });

  const stringifiedParams = normalizedParams.toString();

  return stringifiedParams.length > 0 ? `/api/transactions?${stringifiedParams}` : `/api/transactions`
}

/**
 * @summary List the current user's transactions, most recent first
 */
export const listTransactions = async (params?: ListTransactionsParams, options?: RequestInit): Promise<Transaction[]> => {

  return customFetch<Transaction[]>(getListTransactionsUrl(params),
  {
    ...options,
    method: 'GET'


  }
);}





export const getListTransactionsQueryKey = (params?: ListTransactionsParams,) => {
    return [
    `/api/transactions`, ...(params ? [params] : [])
    ] as const;
    }


export const getListTransactionsQueryOptions = <TData = Awaited<ReturnType<typeof listTransactions>>, TError = ErrorType<ErrorResponse>>(params?: ListTransactionsParams, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof listTransactions>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getListTransactionsQueryKey(params);



    const queryFn: QueryFunction<Awaited<ReturnType<typeof listTransactions>>> = ({ signal }) => listTransactions(params, { signal, ...requestOptions });





   return  { queryKey, queryFn, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof listTransactions>>, TError, TData> & { queryKey: QueryKey }
}

export type ListTransactionsQueryResult = NonNullable<Awaited<ReturnType<typeof listTransactions>>>
export type ListTransactionsQueryError = ErrorType<ErrorResponse>


/**
 * @summary List the current user's transactions, most recent first
 */

export function useListTransactions<TData = Awaited<ReturnType<typeof listTransactions>>, TError = ErrorType<ErrorResponse>>(
 params?: ListTransactionsParams, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof listTransactions>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getListTransactionsQueryOptions(params,options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getGetTransactionUrl = (id: number,) => {




  return `/api/transactions/${id}`
}

/**
 * @summary Get a single transaction by id
 */
export const getTransaction = async (id: number, options?: RequestInit): Promise<Transaction> => {

  return customFetch<Transaction>(getGetTransactionUrl(id),
  {
    ...options,
    method: 'GET'


  }
);}





export const getGetTransactionQueryKey = (id: number,) => {
    return [
    `/api/transactions/${id}`
    ] as const;
    }


export const getGetTransactionQueryOptions = <TData = Awaited<ReturnType<typeof getTransaction>>, TError = ErrorType<ErrorResponse>>(id: number, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof getTransaction>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getGetTransactionQueryKey(id);



    const queryFn: QueryFunction<Awaited<ReturnType<typeof getTransaction>>> = ({ signal }) => getTransaction(id, { signal, ...requestOptions });





   return  { queryKey, queryFn, enabled: id !== null && id !== undefined, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof getTransaction>>, TError, TData> & { queryKey: QueryKey }
}

export type GetTransactionQueryResult = NonNullable<Awaited<ReturnType<typeof getTransaction>>>
export type GetTransactionQueryError = ErrorType<ErrorResponse>


/**
 * @summary Get a single transaction by id
 */

export function useGetTransaction<TData = Awaited<ReturnType<typeof getTransaction>>, TError = ErrorType<ErrorResponse>>(
 id: number, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof getTransaction>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getGetTransactionQueryOptions(id,options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getCreateTransferUrl = () => {




  return `/api/transfers`
}

/**
 * @summary Send money to another user by phone number
 */
export const createTransfer = async (transferInput: TransferInput, options?: RequestInit): Promise<WalletMutationResult> => {

  return customFetch<WalletMutationResult>(getCreateTransferUrl(),
  {
    ...options,
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(transferInput)
  }
);}





export const getCreateTransferMutationOptions = <TError = ErrorType<ErrorResponse>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof createTransfer>>, TError,{data: BodyType<TransferInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof createTransfer>>, TError,{data: BodyType<TransferInput>}, TContext> => {

const mutationKey = ['createTransfer'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof createTransfer>>, {data: BodyType<TransferInput>}> = (props) => {
          const {data} = props ?? {};

          return  createTransfer(data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type CreateTransferMutationResult = NonNullable<Awaited<ReturnType<typeof createTransfer>>>
    export type CreateTransferMutationBody = BodyType<TransferInput>
    export type CreateTransferMutationError = ErrorType<ErrorResponse>

    /**
 * @summary Send money to another user by phone number
 */
export const useCreateTransfer = <TError = ErrorType<ErrorResponse>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof createTransfer>>, TError,{data: BodyType<TransferInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof createTransfer>>,
        TError,
        {data: BodyType<TransferInput>},
        TContext
      > => {
      return useMutation(getCreateTransferMutationOptions(options));
    }

export const getLookupUserUrl = (params: LookupUserParams,) => {
  const normalizedParams = new URLSearchParams();

  Object.entries(params || {}).forEach(([key, value]) => {

    if (value !== undefined) {
      normalizedParams.append(key, value === null ? 'null' : String(value))
    }
  });

  const stringifiedParams = normalizedParams.toString();

  return stringifiedParams.length > 0 ? `/api/users/lookup?${stringifiedParams}` : `/api/users/lookup`
}

/**
 * @summary Look up a user's public profile by phone number
 */
export const lookupUser = async (params: LookupUserParams, options?: RequestInit): Promise<PublicUser> => {

  return customFetch<PublicUser>(getLookupUserUrl(params),
  {
    ...options,
    method: 'GET'


  }
);}





export const getLookupUserQueryKey = (params?: LookupUserParams,) => {
    return [
    `/api/users/lookup`, ...(params ? [params] : [])
    ] as const;
    }


export const getLookupUserQueryOptions = <TData = Awaited<ReturnType<typeof lookupUser>>, TError = ErrorType<ErrorResponse>>(params: LookupUserParams, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof lookupUser>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getLookupUserQueryKey(params);



    const queryFn: QueryFunction<Awaited<ReturnType<typeof lookupUser>>> = ({ signal }) => lookupUser(params, { signal, ...requestOptions });





   return  { queryKey, queryFn, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof lookupUser>>, TError, TData> & { queryKey: QueryKey }
}

export type LookupUserQueryResult = NonNullable<Awaited<ReturnType<typeof lookupUser>>>
export type LookupUserQueryError = ErrorType<ErrorResponse>


/**
 * @summary Look up a user's public profile by phone number
 */

export function useLookupUser<TData = Awaited<ReturnType<typeof lookupUser>>, TError = ErrorType<ErrorResponse>>(
 params: LookupUserParams, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof lookupUser>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getLookupUserQueryOptions(params,options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getListContactsUrl = () => {




  return `/api/contacts`
}

/**
 * @summary List saved payees for the current user
 */
export const listContacts = async ( options?: RequestInit): Promise<Contact[]> => {

  return customFetch<Contact[]>(getListContactsUrl(),
  {
    ...options,
    method: 'GET'


  }
);}





export const getListContactsQueryKey = () => {
    return [
    `/api/contacts`
    ] as const;
    }


export const getListContactsQueryOptions = <TData = Awaited<ReturnType<typeof listContacts>>, TError = ErrorType<ErrorResponse>>( options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof listContacts>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getListContactsQueryKey();



    const queryFn: QueryFunction<Awaited<ReturnType<typeof listContacts>>> = ({ signal }) => listContacts({ signal, ...requestOptions });





   return  { queryKey, queryFn, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof listContacts>>, TError, TData> & { queryKey: QueryKey }
}

export type ListContactsQueryResult = NonNullable<Awaited<ReturnType<typeof listContacts>>>
export type ListContactsQueryError = ErrorType<ErrorResponse>


/**
 * @summary List saved payees for the current user
 */

export function useListContacts<TData = Awaited<ReturnType<typeof listContacts>>, TError = ErrorType<ErrorResponse>>(
  options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof listContacts>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getListContactsQueryOptions(options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getCreateContactUrl = () => {




  return `/api/contacts`
}

/**
 * @summary Save a new payee
 */
export const createContact = async (contactInput: ContactInput, options?: RequestInit): Promise<Contact> => {

  return customFetch<Contact>(getCreateContactUrl(),
  {
    ...options,
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(contactInput)
  }
);}





export const getCreateContactMutationOptions = <TError = ErrorType<ErrorResponse>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof createContact>>, TError,{data: BodyType<ContactInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof createContact>>, TError,{data: BodyType<ContactInput>}, TContext> => {

const mutationKey = ['createContact'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof createContact>>, {data: BodyType<ContactInput>}> = (props) => {
          const {data} = props ?? {};

          return  createContact(data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type CreateContactMutationResult = NonNullable<Awaited<ReturnType<typeof createContact>>>
    export type CreateContactMutationBody = BodyType<ContactInput>
    export type CreateContactMutationError = ErrorType<ErrorResponse>

    /**
 * @summary Save a new payee
 */
export const useCreateContact = <TError = ErrorType<ErrorResponse>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof createContact>>, TError,{data: BodyType<ContactInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof createContact>>,
        TError,
        {data: BodyType<ContactInput>},
        TContext
      > => {
      return useMutation(getCreateContactMutationOptions(options));
    }

export const getDeleteContactUrl = (id: number,) => {




  return `/api/contacts/${id}`
}

/**
 * @summary Remove a saved payee
 */
export const deleteContact = async (id: number, options?: RequestInit): Promise<void> => {

  return customFetch<void>(getDeleteContactUrl(id),
  {
    ...options,
    method: 'DELETE'


  }
);}





export const getDeleteContactMutationOptions = <TError = ErrorType<ErrorResponse>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof deleteContact>>, TError,{id: number}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof deleteContact>>, TError,{id: number}, TContext> => {

const mutationKey = ['deleteContact'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof deleteContact>>, {id: number}> = (props) => {
          const {id} = props ?? {};

          return  deleteContact(id,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type DeleteContactMutationResult = NonNullable<Awaited<ReturnType<typeof deleteContact>>>

    export type DeleteContactMutationError = ErrorType<ErrorResponse>

    /**
 * @summary Remove a saved payee
 */
export const useDeleteContact = <TError = ErrorType<ErrorResponse>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof deleteContact>>, TError,{id: number}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof deleteContact>>,
        TError,
        {id: number},
        TContext
      > => {
      return useMutation(getDeleteContactMutationOptions(options));
    }

export const getListBillersUrl = (params?: ListBillersParams,) => {
  const normalizedParams = new URLSearchParams();

  Object.entries(params || {}).forEach(([key, value]) => {

    if (value !== undefined) {
      normalizedParams.append(key, value === null ? 'null' : String(value))
    }
  });

  const stringifiedParams = normalizedParams.toString();

  return stringifiedParams.length > 0 ? `/api/billers?${stringifiedParams}` : `/api/billers`
}

/**
 * @summary List available recharge and bill payment providers
 */
export const listBillers = async (params?: ListBillersParams, options?: RequestInit): Promise<Biller[]> => {

  return customFetch<Biller[]>(getListBillersUrl(params),
  {
    ...options,
    method: 'GET'


  }
);}





export const getListBillersQueryKey = (params?: ListBillersParams,) => {
    return [
    `/api/billers`, ...(params ? [params] : [])
    ] as const;
    }


export const getListBillersQueryOptions = <TData = Awaited<ReturnType<typeof listBillers>>, TError = ErrorType<unknown>>(params?: ListBillersParams, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof listBillers>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getListBillersQueryKey(params);



    const queryFn: QueryFunction<Awaited<ReturnType<typeof listBillers>>> = ({ signal }) => listBillers(params, { signal, ...requestOptions });





   return  { queryKey, queryFn, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof listBillers>>, TError, TData> & { queryKey: QueryKey }
}

export type ListBillersQueryResult = NonNullable<Awaited<ReturnType<typeof listBillers>>>
export type ListBillersQueryError = ErrorType<unknown>


/**
 * @summary List available recharge and bill payment providers
 */

export function useListBillers<TData = Awaited<ReturnType<typeof listBillers>>, TError = ErrorType<unknown>>(
 params?: ListBillersParams, options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof listBillers>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getListBillersQueryOptions(params,options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}







export const getCreateRechargeUrl = () => {




  return `/api/recharge`
}

/**
 * @summary Pay a recharge or bill using a biller
 */
export const createRecharge = async (rechargeInput: RechargeInput, options?: RequestInit): Promise<WalletMutationResult> => {

  return customFetch<WalletMutationResult>(getCreateRechargeUrl(),
  {
    ...options,
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...options?.headers },
    body: JSON.stringify(rechargeInput)
  }
);}





export const getCreateRechargeMutationOptions = <TError = ErrorType<ErrorResponse>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof createRecharge>>, TError,{data: BodyType<RechargeInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
): UseMutationOptions<Awaited<ReturnType<typeof createRecharge>>, TError,{data: BodyType<RechargeInput>}, TContext> => {

const mutationKey = ['createRecharge'];
const {mutation: mutationOptions, request: requestOptions} = options ?
      options.mutation && 'mutationKey' in options.mutation && options.mutation.mutationKey ?
      options
      : {...options, mutation: {...options.mutation, mutationKey}}
      : {mutation: { mutationKey, }, request: undefined};




      const mutationFn: MutationFunction<Awaited<ReturnType<typeof createRecharge>>, {data: BodyType<RechargeInput>}> = (props) => {
          const {data} = props ?? {};

          return  createRecharge(data,requestOptions)
        }






  return  { mutationFn, ...mutationOptions }}

    export type CreateRechargeMutationResult = NonNullable<Awaited<ReturnType<typeof createRecharge>>>
    export type CreateRechargeMutationBody = BodyType<RechargeInput>
    export type CreateRechargeMutationError = ErrorType<ErrorResponse>

    /**
 * @summary Pay a recharge or bill using a biller
 */
export const useCreateRecharge = <TError = ErrorType<ErrorResponse>,
    TContext = unknown>(options?: { mutation?:UseMutationOptions<Awaited<ReturnType<typeof createRecharge>>, TError,{data: BodyType<RechargeInput>}, TContext>, request?: SecondParameter<typeof customFetch>}
 ): UseMutationResult<
        Awaited<ReturnType<typeof createRecharge>>,
        TError,
        {data: BodyType<RechargeInput>},
        TContext
      > => {
      return useMutation(getCreateRechargeMutationOptions(options));
    }

export const getGetDashboardSummaryUrl = () => {




  return `/api/dashboard/summary`
}

/**
 * @summary Aggregated balance, recent activity, and spend breakdown
 */
export const getDashboardSummary = async ( options?: RequestInit): Promise<DashboardSummary> => {

  return customFetch<DashboardSummary>(getGetDashboardSummaryUrl(),
  {
    ...options,
    method: 'GET'


  }
);}





export const getGetDashboardSummaryQueryKey = () => {
    return [
    `/api/dashboard/summary`
    ] as const;
    }


export const getGetDashboardSummaryQueryOptions = <TData = Awaited<ReturnType<typeof getDashboardSummary>>, TError = ErrorType<ErrorResponse>>( options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof getDashboardSummary>>, TError, TData>, request?: SecondParameter<typeof customFetch>}
) => {

const {query: queryOptions, request: requestOptions} = options ?? {};

  const queryKey =  queryOptions?.queryKey ?? getGetDashboardSummaryQueryKey();



    const queryFn: QueryFunction<Awaited<ReturnType<typeof getDashboardSummary>>> = ({ signal }) => getDashboardSummary({ signal, ...requestOptions });





   return  { queryKey, queryFn, ...queryOptions} as UseQueryOptions<Awaited<ReturnType<typeof getDashboardSummary>>, TError, TData> & { queryKey: QueryKey }
}

export type GetDashboardSummaryQueryResult = NonNullable<Awaited<ReturnType<typeof getDashboardSummary>>>
export type GetDashboardSummaryQueryError = ErrorType<ErrorResponse>


/**
 * @summary Aggregated balance, recent activity, and spend breakdown
 */

export function useGetDashboardSummary<TData = Awaited<ReturnType<typeof getDashboardSummary>>, TError = ErrorType<ErrorResponse>>(
  options?: { query?:UseQueryOptions<Awaited<ReturnType<typeof getDashboardSummary>>, TError, TData>, request?: SecondParameter<typeof customFetch>}

 ):  UseQueryResult<TData, TError> & { queryKey: QueryKey } {

  const queryOptions = getGetDashboardSummaryQueryOptions(options)

  const query = useQuery(queryOptions) as  UseQueryResult<TData, TError> & { queryKey: QueryKey };

  return withQueryKey(query, queryOptions.queryKey);
}
